src/wp-includes/class-wp-scripts.php

```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
+	 * @param array<string, true> $checked  Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return string[] A list of eligible loading strategies that could be used.
 	 */
-	private function get_eligible_loading_strategy( $handle ) {
+	private function filter_eligible_strategies( $handle, $eligible = null, $checked = array() ) {
+		// If no strategies are being passed, all strategies are eligible.
+		if ( null === $eligible ) {
+			$eligible = $this->delayed_strategies;
+		}
+
+		// If this handle was already checked, return early.
+		if ( isset( $checked[ $handle ] ) ) {
+			return $eligible;
+		}
+
+		// Mark this handle as checked.
+		$checked[ $handle ] = true;
+
+		// If this handle isn't registered, don't filter anything and return.
 		if ( ! isset( $this->registered[ $handle ] ) ) {
-			return '';
+			return $eligible;
+		}
+
+		// If the handle is not enqueued, don't filter anything and return.
+		if ( ! $this->query( $handle, 'enqueued' ) ) {
+			return $eligible;
 		}
 
+		$is_alias = (bool) ! $this->registered[ $handle ]->src;
 		$intended_strategy = $this->get_data( $handle, 'strategy' );
 
-		/*
-		 * Handle known blocking strategy scenarios.
-		 * - An empty strategy is synonymous with blocking.
-		 * - A script alias (where $src is false) must always be blocking since the after inline script cannot be
-		 *   delayed as there is no external script tag and thus no load event at which the inline script can be run.
-		 */
-		if ( empty( $intended_strategy ) ) {
-			return '';
+		// For non-alias handles, an empty intended strategy filters all strategies.
+		if ( ! $is_alias && empty( $intended_strategy ) ) {
+			return array();
 		}
 
-		// Aliases that have before/after inline scripts can never be delayed since there is no load event.
-		if ( ! $this->registered[ $handle ]->src && $this->has_inline_script( $handle ) ) {
-			return '';
+		// Handles with inline scripts attached in the 'after' position cannot be delayed.
+		if ( $this->has_inline_script( $handle, 'after' ) ) {
+			return array();
 		}
 
-		// Handle async strategy scenario.
-		if ( 'async' === $intended_strategy && $this->has_only_delayed_dependents( $handle, true ) ) {
-			return 'async';
+		// If the intended strategy is 'defer', filter out 'async'.
+		if ( $intended_strategy === 'defer' ) {
+			$eligible =  array( 'defer' );
 		}
 
-		// Handling defer strategy scenarios. (The intended strategy could actually be 'async' here.)
-		if ( $this->has_only_delayed_dependents( $handle ) ) {
-			return 'defer';
+		$dependents = $this->get_dependents( $handle );
+
+		// Recursively filter eligible strategies for dependents.
+		foreach ( $dependents as $dependent ) {
+			// Bail early once we know the eligible strategy is blocking.
+			if ( empty( $eligible ) ) {
+				return array();
+			}
+
+			$eligible = $this->filter_eligible_strategies( $dependent, $eligible, $checked );
 		}
```

Comment:
I'm not completely sure this logic is correct. The way it is written it always returns the eligible strategies for the last dependent, unless any of the dependents returned an empty array.

This may not be a common problem in practice, but I think it is possible for one dependent to e.g. return `array( 'defer' )` and another dependent later to return `array( 'async' )` or `array( 'async', 'defer' )`. Shouldn't the return value then be the stricter one, i.e. `array( 'defer' )`?

Reply:
We’re passing the array of eligible strategies to this function, which then gets filtered by the characteristics of the dependent, so the filtering is cumulative.

Reply:
Thanks for clarifying @joemcgill, that makes sense. Just a recommendation then from my end would be to make `$eligible` the first parameter of that method, as that makes it more intuitive that that value is what's being filtered in the method.

Reply:
The only required param in this method is the first one, `$handle`, so I think it's best to leave that in the first position.